HPCC-27122 Create Test Auth secmgr for ECLWatch

wangkx opened the pull request:
1. Add security plugin: testauthSecurity.
2. Use the security plugin as default auth for ECLWatch.
3. Read the settings for the security plugin.
4. Change AuthDomains type based on the settings.

Will create another PR to configure more AuthDomains settings.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-187.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: d39dc59a6f59b011e1f2a4997d11749758d6175f
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.6.13-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  401 sec (00:06:41) |  100 sec (00:01:40) |   18 sec (00:00:18) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  538 sec (00:08:58) |

RussWhitehead reviewed (changes requested):
A couple of questions for @afishbeck and @wangkx 
1) Why is this necessary?  Seems like a risk if users mistakenly try to instantiate it
2) Will this be documented?  Seems like an ESP developer-only tool?
2) Shouldn't this be part of the LN branch?   I don't like that we are making it available to open source users

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        const char* xpath = nodes.item(0);
+        if (fileScopeAccesses->hasProp(xpath))
+        {
+            OWARNLOG("Duplicated accesss settting for scope %s.", scope);
```
Will fix

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+    {
+        if (!secMgrCfg || !bindConfig)
+            return;
+        readUserAccess(secMgrCfg, bindConfig);
```
The readUserAccess() is called when an ESP is starting. If an exception is thrown within, the ESP will be failed to start.

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        StringBuffer userN, password;
+        userN.set(user.getName());
+        if (userN.isEmpty())
+			return false;
```
Will fix

richardkchapman commented:
@wangkx This seems to have stalled?

HPCCSmoketest commented:
Process of PR-15919, label: h27122 starts now.
The reason of this test is: Forced to re-test.
Commit ID: c21bb065b03b81a2982c99069f256d85f1b2e6fe
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-181.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

afishbeck commented:
@richardkchapman @ghalliday this feature is in response to several discussions about having a way of testing authorization and the basic functionality of logging in to EclWatch without needing LDAP or other back end security system.

Could you please take a look with an eye on whether this meets your expectations?  Review at least the high level design, and the way it would be customized via values.yaml?  And also any suggestions about better names would be appreciated.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-181.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: c21bb065b03b81a2982c99069f256d85f1b2e6fe
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.6.13-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  398 sec (00:06:38) |   96 sec (00:01:36) |   17 sec (00:00:17) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  530 sec (00:08:50) |

richardkchapman commented:
Looks like vcpkg has been updated, probably not wanted here

wangkx commented:
Not sure why the vcpkg got updated. But, I shall fix it.

wangkx committed:
HPCC-27122 Create Test Auth secmgr for ECLWatch

1. Add security plugin: testauthSecurity.
2. Use the security plugin as default auth for ECLWatch.
3. Read the settings for the security plugin.
4. Change AuthDomains type based on the settings.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented on helm/hpcc/values.yaml:
```diff
@@ -493,7 +497,7 @@ esp:
   ## Pre-configured esp applications include eclwatch, eclservices, and eclqueries
   application: eclwatch
-  auth: none
+  auth: test
```
Any other opinion?

RussWhitehead replied:
IMHO, it will be frustrating to first time users that download and install, then try to load ECLWatch.  They will be forced to enter a username/pwd, which they have to retrieve from the docs (and possibly provides a false sense of security).  Just one more frustrating step in the learning curve, and I am not sure how it benefits a user that isn't a tester

afishbeck replied:
I probably wouldn't make it the default, unless @richardkchapman or @ghalliday disagree.  I know there was a discussion similar to making it the default at some point.  But we would need to make sure the user knew what to put in as the username if we did.

ghalliday replied:
I think it would make sense for the startall.sh to default it to test.  I think we should have a separate developer.yaml file which is used by the startall.sh to ensure certain components are configured how we would expect.  A separate PR though.

wangkx replied:
Changing to none

HPCCSmoketest commented:
Process of PR-15919, label: h27122 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 717647dd5fedfb3d1b30de7e9e1b8b083045a3ab
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-67.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-67.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 717647dd5fedfb3d1b30de7e9e1b8b083045a3ab
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.6.17-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  399 sec (00:06:39) |   95 sec (00:01:35) |   17 sec (00:00:17) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  526 sec (00:08:46) |

afishbeck commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+
+            StringBuffer userName, password;
+            if (!readUserNamePassword(userSettings, userName, password))
+                continue;
```
It's up to you, a bad config can confuse things especially while using this for something like a regression test.. so stopping ESP can be good to prevent confusion.

wangkx replied:
Changing to throw exception

afishbeck commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+            getSecretKeyValue(userName, secretTree, "username");
+            getSecretKeyValue(password, secretTree, "password");
+            if (userName.isEmpty() || password.isEmpty())
+                throw makeStringExceptionV(-1, "Error retrieving userName/password for %s.", secretKey);
```
To answer your quetion about "Does a valid secret always have both username and password?".
You get to decide what is mandatory in your secrets.

wangkx replied:
Changing to not read username from secret.

ghalliday commented on esp/applications/eclwatch/testauth_authorization_map.yaml:
```diff
@@ -0,0 +1,43 @@
+         #secretkey: testauthusercreds1     # optional
+         #vaultkey: testauthuservaultcreds1 # optional
+         defaults:
+            filescope: FULL
```
minor: lower case for full and none would be more common in a yaml file.

wangkx replied:
Will fix

ghalliday commented on esp/applications/eclwatch/testauth_authorization_map.yaml:
```diff
@@ -0,0 +1,43 @@
+         #secretkey: testauthusercreds2     # optional
+         #vaultkey: testauthuservaultcreds2 # optional
+         defaults:
+            filescope: NONE
```
minor: The capitalisation isn't very consistent.  E.g., LibName is camel case with leading upper.  useResourceMapsFrom is camel with leading lower.  authtype, filescope etc. is all lower (rather than authType, fileScope)
Most entries in the values.yaml use camel case with leading lower.

wangkx replied:
Some of them are based on the existing code and environment.xml. For example, system\security\shared\secloader.hpp(49): secMgrCfg->getProp("@LibName", libName);. We cannot change the environment.xml. Not sure we want to change the existing code to handle both cases. I will check them one by one to make it consistent as much as I can.

afishbeck replied:
Right.  If we change names that are currently a pass through to match old config then we need to create a mapping.  We've avoided that, but could do more of it if it's worth it.  If we do much more of it we can create a mapping file rather than mapping in c++.

ghalliday commented on esp/bindings/http/platform/httpbinding.cpp:
```diff
@@ -98,6 +98,21 @@ static IXmlSchema* createXmlSchema(const char* schema)
+
+    const char* srvType = srvCfg->queryProp("@type");
+    if (!isEmptyString(srvType))
+        bndCfg->addProp("@serviceType", srvType);
```
trivial: setProp is more natural since you can't have more than one instance of an attribute.

wangkx replied:
Will change

ghalliday commented on esp/platform/application_config.cpp:
```diff
@@ -89,6 +89,21 @@ static void copyDirectories(IPropertyTree *target, IPropertyTree *src)
     }
 }
 
+void addAuthDomains(const char *authtype, IPropertyTree *legacyEsp)
```
picky: authType would be the normal casing.

wangkx replied:
Will fix

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        const char* secretKey = userSettings.queryProp("@secretkey");
+        if (!isEmptyString(secretKey))
+        {
+            const char* vaultKey = userSettings.queryProp("@vaultkey");
```
question: Why does this need a separate vault secret?  @afishbeck am I correct the same key can be used for both, when it is passed to getSecret().

afishbeck replied:
The name is actually wrong, it should just be vault.  If there are more than one vaults configured (or a mixture of vault and Kubernetes secrets) the vault name would specify which one to look for the secret in.

The vault name is being passed to the right place below, so the attribute just needs to be renamed.

ghalliday replied:
It should also be made clear that the vault property restricts the secret to be resolved in that valut.  I think the default behaviour is to search all vaults.  @afishbeck Is this really needed?

wangkx replied:
Removing vault related code

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+            if (!secretTree)
+                throw makeStringExceptionV(-1, "Error retrieving the secret for %s.", secretKey);
+
+            getSecretKeyValue(userName, secretTree, "username");
```
question: @afishbeck would you expect the username to be in the secret?  I was expecting only the password.

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        if (isEmptyString(userName))
+            return false;
+
+        const char* passwordStr = userSettings.queryProp("@password");
```
This isn't commented in the values.yaml

wangkx replied:
Will do

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+    StringAttr userName, password;
+    SecAccessFlags defaulteECLWUScopeAccess = SecAccess_Unavailable;
+    SecAccessFlags defaultFileScopeAccess = SecAccess_Unavailable;
+    SecAccessFlags defaultFeaturAccess = SecAccess_Unavailable;
```
typo? Featur->Feature

wangkx replied:
Will fix

ghalliday reviewed (commented):
Some initial comments - continuing tomorrow.

wangkx commented:
Ok, I will retarget this to master.

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        Owned<IPropertyTreeIterator> iter = authconfig->getElements(".//Feature");
+        ForEach(*iter)
+        {
+            IPropertyTree& feature = iter->query();
```
similar comment about shared code in the base class.

wangkx replied:
Will do

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        SecAccessFlags  resultFlag = SecAccess_Unavailable;
+        try
+        {
+            if (!authenticate(user))
```
Shouldn't the authentication be done as a separate stage to authorization?

wangkx replied:
The logic here matches the existing CLdapSecManager::authorizeEx().

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        try
+        {
+            if (!authenticate(user))
+                return resultFlag;
```
minor: Clearer when reading/reviewing the code if this (and line 423) return SecAccess_Unavailable.

wangkx replied:
Will change

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        return streq(password.str(), userAccessPtr->queryPassword());
+    }
+
+    bool authorize(ISecUser& user, ISecResourceList* Resources, IEspSecureContext* secureContext = nullptr) override
```
minor: I think this is public in the base interface, so not sure it makes sense for it to be protected (and other functions)

wangkx replied:
Will change to public.

ghalliday reviewed (commented):
@wangkx a few fairly minor comments.  Some changes may be outside the scope of this PR.

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+            }
+            p += 2;
+        }
+        if (fullPathOnly && (scope.charAt(scope.length() - 1) == '/'))
```
Will check

ghalliday replied:
It may not be possible to provoke if the caller prevents it being called with a blank string.

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        if (element->hasProp("@access"))
+        {
+            int newAccess = element->getPropInt("@access");
+            if (newAccess < SecAccess_Read)
```
This behaviour matches the existing security in bare metal environment (please see the comments in system\security\shared\caching.cpp, line 541-550). @RussWhitehead please comment.

afishbeck replied:
I had the same initial thought about how it should work.  But we thought we probably wanted to match ldap for testing purposes?  Open to suggestions based on the intended use cases.  When we create a new security manager we could change that (OAUTH2?) and might need a flavor of test security that matches that behavior.

RussWhitehead replied:
Any DENY specified in a filescope node applies to that node and all children. So if we have
scope1::scope2::scope3 with a DENY at scope2, then scope3 will also be denied.

ghalliday replied:
I understand that is the behaviour - but it seems strange (and inconsistent with allowing deny all as the default).  However discussion is outside the scope of this PR.

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+    SecAccessFlags getAccessFlagsEx(SecResourceType rtype, ISecUser& user, const char* resourceName, IEspSecureContext* secureContext = nullptr) override
+    {
+        if (isEmptyString(resourceName))
+            return SecAccess_Unavailable;
```
The logic here and line 525 match with CLdapSecManager::getAccessFlagsEx() and CLdapSecManager::authorizeEx().

ghalliday commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+                                                          SMF_AuthorizeFileScope_Named | SMF_AuthorizeWorkUnitScope_Named;
+    static const SecFeatureSet s_safeFeatureMask = s_implementedFeatureMask | SMF_CreateSettingMap;
+
+    MapStringTo<CUserAccess*> userAccessMap;
```
The problem is that kill will not free up the items either.  Because they are mapped to pointers rather than link counted pointers

wangkx replied:
I saw the following code in dfuxreflib.cpp.
  manager.logicalnamemap.setValue(name.str(), lnentry);
  lnentry->Link();
  manager.logicalnamelist.append(*lnentry);
Will it free up the items if I do the similar here? Any other suggestion?

ghalliday replied:
Appending them to a list would be one way of cleaning them up.

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+    SecAccessFlags authorizeEx(SecResourceType rtype, ISecUser& user, const char* resourceName, IEspSecureContext* secureContext) override
+    {
+        if (isEmptyString(resourceName))
+            return SecAccess_Full; //Copy from LdapSecurity lib. Not sure this is safe.
```
The logic here matches with 
CLdapSecManager::authorizeEx(SecResourceType rtype, ISecUser & user, const char * resourcename, IEspSecureContext* secureContext).

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+            if (!userAccess)
+                return resultFlag;
+
+            CUserAccessPtr userAccessPtr = *userAccess;
```
Will fix

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+            return false;
+
+        CUserAccessPtr userAccessPtr = *userAccess;
+        password.set(user.credentials().getPassword());
```
It is to avoid checking for null. But, I can add a check to avoid the copy.

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+    void addFeatureAccess(const char* resource, const char* access)
+    {
+        SecAccessFlags flag = getSecAccessFlagValue(access);
+        if (!featureAccesses.getValue(resource))
```
It should be OK if just overwriting.

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+
+    bool logoutUser(ISecUser& user, IEspSecureContext* secureContext = nullptr) override
+    {
+        user.setAuthenticateStatus(AS_UNKNOWN);
```
Will do

wangkx commented on system/security/plugins/testauthSecurity/testauthSecurity.cpp:
```diff
@@ -0,0 +1,591 @@
+        Owned<IPropertyTreeIterator> iter = authconfig->getElements(".//Location");
+        ForEach(*iter)
+        {
+            IPropertyTree& location = iter->query();
```
Will do

HPCCSmoketest commented:
Process of PR-15919, label: h27122 starts now.
The reason of this test is: Base branch changed.
Commit ID: b200ee5b5067ded057197a70c92434fcbef521f1
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-44.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-44.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: b200ee5b5067ded057197a70c92434fcbef521f1
Containerized:True
Number of Error(s): 3
make[2]: *** [esp/src/CMakeFiles/eclwatch_fetchdeps.dir/build.make:73: esp/src/lib/package-lock.json] Error 1
make[1]: *** [CMakeFiles/Makefile2:11451: esp/src/CMakeFiles/eclwatch_fetchdeps.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2022-04-18_15-36-27
time logs started
Prep time   :    17 sec (00:00:17)
Build time  :   145 sec (00:02:25)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   162 sec (00:02:42)
time logs end

[Build error:](https://gist.github.com/75f460ebf6363c1a3442cd2ff8b63689)
Build log [Build log](https://gist.github.com/HPCCSmoketest/75f460ebf6363c1a3442cd2ff8b63689#file-result-22-04-18-15-33-27-log)

HPCCSmoketest commented:
Process of PR-15919, label: h27122 starts now.
The reason of this test is: Forced to re-test.
Commit ID: d4156e3d63b9e624bd1f8f6a2fc3e2b99b0e63e2
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-164.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-164.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: d4156e3d63b9e624bd1f8f6a2fc3e2b99b0e63e2
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.7.0-trunk0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   23 sec (00:00:23) |  399 sec (00:06:39) |   96 sec (00:01:36) |   18 sec (00:00:18) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  536 sec (00:08:56) |

HPCCSmoketest commented:
Process of PR-15919, label: h27122 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 1cdcfd1d58b18e6da5479a6e17a0f43cd43a67f1
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-219.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

wangkx commented:
@ghalliday please review my last 3 commits.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-219.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: 1cdcfd1d58b18e6da5479a6e17a0f43cd43a67f1
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.7.0-trunk0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   23 sec (00:00:23) |  399 sec (00:06:39) |   97 sec (00:01:37) |   18 sec (00:00:18) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  537 sec (00:08:57) |

ghalliday reviewed (approved):
@wangkx one minor comment, but looks much cleaner.

HPCCSmoketest commented:
Process of PR-15919, label: h27122 starts now.
The reason of this test is: Forced to re-test.
Commit ID: e5b1843c017b88b3e8aa3e4d58fbdb36429d7736
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-235.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-235.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: e5b1843c017b88b3e8aa3e4d58fbdb36429d7736
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.7.0-trunk0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  397 sec (00:06:37) |   96 sec (00:01:36) |   18 sec (00:00:18) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  528 sec (00:08:48) |

RussWhitehead commented:
@wangkx Looks good

RussWhitehead reviewed (approved):
Looks good

richardkchapman commented:
@wangkx Please squash

HPCCSmoketest commented:
Process of PR-15919, label: h27122 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 1ff14f31fc3835611779497e00e86253345bd6e8
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-46.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Build only for containerized environment.
